pkg/pool-weighted/contracts/smart/InvestmentPool.sol

```diff
@@ -0,0 +1,271 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/ReentrancyGuard.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/WordCodec.sol";
+
+import "../BaseWeightedPool.sol";
+import "./WeightCompression.sol";
+
+/**
+ * @dev Weighted Pool with mutable weights, designed to support V2 Liquidity Bootstrapping
+ */
+contract InvestmentPool is BaseWeightedPool, ReentrancyGuard {
+    // The Pause Window and Buffer Period are timestamp-based: they should not be relied upon for sub-minute accuracy.
+    // solhint-disable not-rely-on-time
+
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+    using WeightCompression for uint256;
+
+    // State variables
+
+    // The current number of tokens in the pool
+    // Technically redundant; cached here to avoid calling getTokens on the pool,
```

Comment:
Why do you say this is redundant? This seems perfectly fine to me.

Reply:
I mean the ground truth is the Vault, and it could be slightly confusing to read from a cache (that could technically get out of sync).

Maybe I'll remove the word "redundant" and explicitly call it a cache; that should improve the clarity.

Reply:
But we have `_totalTokens` in all pools, and the Vault is nonetheless the source of ground truth in those pools. Does the fact that it's mutable here actually make a difference to that design pattern?

Reply:
Well it's going away in the next PR anyway (moved into packed storage), so it doesn't really matter.